Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp

```diff
@@ -33,43 +40,150 @@ static std::string GetJoystickName(int index)
+}
+
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+static Common::Event s_init_event;
+static Uint32 s_stop_event_type;
+static Uint32 s_populate_event_type;
+static Common::Event s_populated_event;
+static std::thread s_hotplug_thread;
+
+static bool HandleEventAndContinue(SDL_Event& e)
+{
+  if (e.type == SDL_JOYDEVICEADDED)
   {
-    // Failed to initialize
-    return;
+    OpenAndAddDevice(e.jdevice.which);
+    g_controller_interface.InvokeHotplugCallbacks();
   }
+  else if (e.type == SDL_JOYDEVICEREMOVED)
+  {
+    g_controller_interface.RemoveDevice([&e](const auto& device) {
+      const Joystick* joystick = dynamic_cast<const Joystick*>(device);
+      return joystick && SDL_JoystickInstanceID(joystick->GetSDLJoystick()) == e.jdevice.which;
+    });
+    g_controller_interface.InvokeHotplugCallbacks();
+  }
+  else if (e.type == s_populate_event_type)
+  {
+    for (int i = 0; i < SDL_NumJoysticks(); ++i)
+      OpenAndAddDevice(i);
+    s_populated_event.Set();
+  }
+  else if (e.type == s_stop_event_type)
+  {
+    return false;
+  }
+
+  return true;
+}
+#endif
+
+void Init()
+{
+#if !SDL_VERSION_ATLEAST(2, 0, 0)
+  if (SDL_Init(SDL_INIT_JOYSTICK) != 0)
+    ERROR_LOG(SERIALINTERFACE, "SDL failed to initialize");
+  return;
+#else
+  s_hotplug_thread = std::thread([] {
+    Common::ScopeGuard quit_guard([] {
+      // TODO: there seems to be some sort of memory leak with SDL, quit isn't freeing everything up
+      SDL_Quit();
+    });
+
+    {
+      Common::ScopeGuard init_guard([] { s_init_event.Set(); });
+
+      if (!SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_HAPTIC))
+      {
+        ERROR_LOG(SERIALINTERFACE, "SDL failed to initialize");
+        return;
+      }
+
+      Uint32 custom_events_start = SDL_RegisterEvents(2);
+      if (custom_events_start == static_cast<Uint32>(-1))
+      {
+        ERROR_LOG(SERIALINTERFACE, "SDL failed to register custom events");
+        return;
+      }
+      s_stop_event_type = custom_events_start;
+      s_populate_event_type = custom_events_start + 1;
+
+      // Drain all of the events and add the initial joysticks before returning. Otherwise, the
+      // individual joystick events as well as the custom populate event will be handled _after_
+      // ControllerInterface::Init/RefreshDevices has cleared its list of devices, resulting in
+      // duplicate devices.
+      SDL_Event e;
+      while (SDL_PollEvent(&e) != 0)
+      {
+        if (!HandleEventAndContinue(e))
+          return;
+      }
+    }
+
+    SDL_Event e;
+    while (SDL_WaitEvent(&e) != 0)
+    {
+      if (!HandleEventAndContinue(e))
+        return;
+    }
+  });
+
+  s_init_event.Wait();
+#endif
+}
+
+void DeInit()
+{
+#if !SDL_VERSION_ATLEAST(2, 0, 0)
+  SDL_Quit();
+#else
+  if (!s_hotplug_thread.joinable())
+    return;
+
+  SDL_Event stop_event;
+  std::memset(&stop_event, 0, sizeof(stop_event));
+  stop_event.type = s_stop_event_type;
+  SDL_PushEvent(&stop_event);
+
+  s_hotplug_thread.join();
```

Comment:
should this not check if the thread is joinable first before attempting to join it?

Reply:
It does, see beginning of function.